src/Utilities/ProcessExtensions.cs

```diff
@@ -0,0 +1,122 @@
+using System.Diagnostics;
+using System.IO;
+using System.Runtime.InteropServices;
+using Microsoft.Build.Shared;
+
+namespace Microsoft.Build.Utilities
+{
+    internal static class ProcessExtensions
+    {
+#if !RUNTIME_TYPE_NETCORE
+        private static readonly bool _isWindows = true;
+#else
+        private static readonly bool _isWindows = RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
+#endif
+
+        public static void KillTree(this Process process, int timeout)
+        {
+            string stdout;
+            if (_isWindows)
+            {
+                try
+                {
+                    // issue the kill command
+                    NativeMethodsShared.KillTree(process.Id);
+                }
+                catch (InvalidOperationException)
+                {
+                    // The process already exited, which is fine,
+                    // just continue.
+                }
+
+                // wait until the process finishes exiting/getting killed. 
+                // We don't want to wait forever here because the task is already supposed to be dieing, we just want to give it long enough
+                // to try and flush what it can and stop. If it cannot do that in a reasonable time frame then we will just ignore it.
+                process.WaitForExit(timeout);
+            }
+            else
+            {
+                var children = new HashSet<int>();
+                GetAllChildIdsUnix(process.Id, children, timeout);
+                foreach (var childId in children)
+                {
+                    KillProcessUnix(childId, timeout);
+                }
+                KillProcessUnix(process.Id, timeout);
+            }
+        }
+
+        private static void GetAllChildIdsUnix(int parentId, ISet<int> children, int timeout)
+        {
+            string stdout;
+            var exitCode = RunProcessAndWaitForExit(
+                "pgrep",
+                $"-P {parentId}",
+                timeout,
+                out stdout);
+
+            if (exitCode == 0 && !string.IsNullOrEmpty(stdout))
+            {
+                using (var reader = new StringReader(stdout))
+                {
+                    while (true)
+                    {
+                        var text = reader.ReadLine();
+                        if (text == null)
+                        {
+                            return;
+                        }
+
+                        int id;
+                        if (int.TryParse(text, out id))
+                        {
+                            children.Add(id);
+                            // Recursively get the children
+                            GetAllChildIdsUnix(id, children, timeout);
+                        }
+                    }
+                }
+            }
+        }
+
+        private static void KillProcessUnix(int processId, int timeout)
+        {
+            string stdout;
+            RunProcessAndWaitForExit(
```

Comment:
Hah, I was just typing up a comment about the dual meanings of `timeout`. I think it'd be better to have a fairly short timeout for the `kill` that's isolated from the "wait X to ensure that everything's done".

For using `signal(3)`, that does seem like a cleaner approach--but will it be as multi-platform-friendly? My years on Windows don't give me a good sense here. I can understand the current maybe-inefficient `kill` code.

---
In reply to: [74337936](https://github.com/Microsoft/msbuild/pull/886#discussion_r74337936) [](ancestors = 74337936)